2 d world

buddyuchiha opened the pull request.

buddyuchiha committed:
added tech

buddyuchiha committed:
added classes such as: base classes (Point, Entity, Creatre, Map, Simulation), static classes (Grass, Rock, Tree), dynamic classes (Herbivore, Predator)

buddyuchiha committed:
beginning Action realiztiond
added:
-CreateAction class with generator methods
edited:
-print_map function in Map class

buddyuchiha committed:
generate static and dynamic objects
added:
- generate object, static and dynamic objects
edited:
- random_cord method, added exception
-  generator logic, we have static coords for random.choice

buddyuchiha committed:
changed hash to point in generate_object

buddyuchiha committed:
new Action init_actions realization
edit:
- deleted unnecesarry functions
-added dict self.capacity
-added the simplest realization for add_entity
-added entety in map

buddyuchiha committed:
beginning move realization

buddyuchiha committed:
added closest grass for herbivore

buddyuchiha committed:
added make move for Herbivore

buddyuchiha committed:
added make_move realiation for Predator
edit:
-move_creature realization for Predator and Herbivore
-isinstances for entity_type in bfs
-make_move for predator
-class Creature relocate to module dynamic

buddyuchiha committed:
new structure of the project
edit:
-created class simulation with necessary subclasses
-deleted unnecessary modules

Drowchik commented on base.py:
```diff
@@ -0,0 +1,27 @@
+from abc import ABC, abstractmethod
+
+class Point():
+    def __init__(self, x = None, y = None):
```
только int

buddyuchiha committed:
refactored function to small functions
edit:
-make move to get_positions, get_goal point

buddyuchiha committed:
strange mistake @_@

buddyuchiha committed:
mistake fixed

buddyuchiha committed:
refactoring

buddyuchiha committed:
refactored

buddyuchiha committed:
complete my realization
add:
-threadings to input and render info
-class Render for render map
-functions in class Simulation to start, stop etc

buddyuchiha committed:
fast pep 8 (•ᴗ•)

Drowchik commented on src/base/base.py:
```diff
@@ -0,0 +1,26 @@
+from abc import ABC
+
+class Point():
+    def __init__(self, x , y):
```
Если это итоговая версия нужны тайп хинты
    def __init__(self, x: int , y: int) -> None:

Drowchik commented on src/entities/dynamic.py:
```diff
@@ -0,0 +1,77 @@
+    def is_dead(self) -> bool:
+        return self.hp <= 0
+    
+    def make_move(self, point: Point, path: list[tuple[int, int]]) -> Point:   
```
а в чем разница с Predator и не совсем понимаю логику метода

buddyuchiha replied:
если is_dead, то я сделал этот метод, чтобы инкапсулировать эту логику, будто потом запарно в других классах и функциях переопределять эту логику по много раз

если make_move, то тут происходит проверка на условие "находится ли Predator рядом с Hebivore" тип если путь == 1, то да

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+
+class Simulation():
+    def __init__(self, size: int):
+        self.map = self.Map(size)
```
я немного не понимаю, почему вы карту и другие классы не передаете из вне? В плане почему инициализируете внутри? 

Пример: мне надо проверить класс Simulation, а он зависит сразу от многих других классов, и я не могу их замокать, так как вы инициализируте внутри, (замокать - подставить аналог класса, без реализации, где я сам говорю что должны методы вернуть) и если буду тестить класс simulation я не пойму это в нем проблема или в зависимых классах

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+            obj.point = goal_point
+            self.map_dict[goal_point] = obj
+         
+        def get_object(self, obj: Herbivore | Predator) -> Grass | Herbivore:
```
а зачем тут self... как вообще это с картой соотносится

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                        self.map.add_entity(point, key)
+                        
+        def turn_actions(self) -> None:
+            temp_map = self.map.map_dict.copy()
```
а зачем здесь копия? ты же никак не меняешь по сути ее

buddyuchiha replied:
При вставке и удалении меняется размер словаря, из-за этого ловим RuntimeError

<img width="891" height="240" alt="image" src="https://github.com/user-attachments/assets/c54603d8-be64-49a5-8e18-e2c72b3e38af" />

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                if(isinstance(value, Creature)):
+                    self.map.move_creature(key, value)
+            
+            for reset_entity in self.reset_entities:
```
я не уверен нужно ли там хранить сущности таким образом reset_entities -> когда захотим менять или добавлять новую сущность тут придется добавлять вспоминать про эту часть кода

buddyuchiha replied:
я это делал опираясь на универсальность, условно в код нужно добавить еще один возрождаемый объект, все что требуется в массив написать название его:

<img width="259" height="73" alt="image" src="https://github.com/user-attachments/assets/737af667-195d-4563-84a2-51a3f5f04e6e" />

далее по циклу уже все подставится:

<img width="535" height="305" alt="image" src="https://github.com/user-attachments/assets/b5a5f6ac-6a14-4362-aa6e-73b2c8bdc8e7" />

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                entity_count = self.map.get_entity_count(reset_entity) 
+                
+                points = self.map.get_available_points()                    
+                if not points:  
```
у тебя в теории из-за того что мы изначально настраиваем заполненность карты определенную, то не может быть, чтобы не было пустых клеток, и я не особо понимаю причину снова находить все пустые точки их же может быть очень много, мы ушли от идеи хранить все поля (как мы отрисовываем карту, а тут мы храним снова отдельно все точки пустые... странно)

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+        
+        def init_actions(self) -> None:
+            for x in range(self.map.size):
+                self.coords = [i for i in range(self.map.size)]
```
зачем именно в self их хранить? почему просто как переменную и передавать?

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+            self.map = map 
+            
+            dynamic_capacity = int(self.map.size * 0.5)
+            static_capacity = int(self.map.size * 0.2)
```
довольно логичное разделение, но нельзя это сразу сделать, в плане без разделения на вот эти части.

то есть к примеру так

    "Grass": 0.1 * self.map.size
    "Rock": 0.1 * self.map.size
    "Tree": 0.1 * self.map.size
    "Herbivore": 0.2 * self.map.size
    "Predator": 0.15 * self.map.size

buddyuchiha replied:
Ну я это сделал опираясь на то, чтобы можно было сразу задавать всем один параметр для вместимости, тип мы меняем один параметр и сразу задаем остальным динамическим объектам вместимость

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                    continue
+                
+                entity_capacity = self.capacity.get(reset_entity) \
+                    * int(self.map.size / 2)
```
для чего вот этот коэфициент? Дополнительно? Просто из логики сверху про динамическое и статическое 

dynamic_capacity = int(self.map.size * 0.5)
 static_capacity = int(self.map.size * 0.2)
            
   а у тебя сущности reset_entities разного типа, тогда и коэфициент должен быть разным?

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                entity_capacity = self.capacity.get(reset_entity) \
+                    * int(self.map.size / 2)
+                while entity_count < entity_capacity and points: 
+                    point = choice(points)
```
def spawn_objects(self) -> None:
        for obj, proportion in self.proportion.items():
            active_obj = self.world.get_all_obj(obj)
            self.add_object(self._get_number_obj(proportion, active_obj), obj)

    def _get_number_obj(self, proportion, active_obj) -> None:
        return (floor(proportion * self.map.get_area())) - len(active_obj)

    def _add_object(self, count, obj) -> None:
        while (count):
            point = Point(random.randint(0, self.world.weight - 1), random.randint(0, self.world.height - 1))
            if not self.world.get_entity(point):
                self.map_matrix.add_object(obj(point))
                count -= 1

посмотри может быть что-то подобное, у тебя конечно по другому реализовано и твой вариант в целом тоже норм, но тут хранишь много пустых точек, что не особо круто =_+ по сути получается тогда ты практически всегда хранишь всю карту entity + пустые, вот _add_object норм, наверное будет, ты просто ставишь сущность, если точка не занята и все.

Насчет пустых точек они всегда должны быть, мы должны задавать такие размеры, чтобы не задумывались (наши коэфициенты порешают), если у тебя они большие понизь, условно чтобы мы поставили ограничения размера карты от 10 на 10, можно больше, меньше нельзя и у тебя все нормально работало

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+                ]     
+            return len(objects)
+        
+        def get_available_points(self) -> list[Point]:
```
я считаю данный метод не нужен

Drowchik commented on src/simulation/simulation.py:
```diff
@@ -0,0 +1,325 @@
+            
+            self.move_entity(point, obj, new_coord)
+        
+        def move_creature(self, point: Point, obj: Creature) -> None:
```
насчет bfs и путей нужно смотреть еще, я пока не успел, но в идеале я хотел бы один move_creature условный, а не  два отдельных move_herbivore и move_predator, надо вместе смотреть, либо я один потыкаюсь

buddyuchiha committed:
making corrections
edit:
- new reset realzation by creating reset_action and deleted get_availbale poins
-edited make_move for better and cleaner code
-get_object now is static method
-now you must create map etc before creatin Simulation class
-reworked random_cord

buddyuchiha committed:
last fixes (＾▽＾)

buddyuchiha committed:
normal architecture (¬‿¬)

buddyuchiha committed:
Create README.md
